### PR TITLE
feat: aptos da client

### PR DESCRIPTION
**Pull request by Tranduy1dol**
# Check List

 - [ ] Don't forget to squash commits into meaningful chunks before merging
 - [ ] Check every test passed.
 - [ ] Did you split imports into std, third parties, and custom parts?
 - [ ] Format your code.
   - [ ] cargo fmt
   - [ ] cargo clippy
   - [ ] `taplo fmt --config taplo/taplo.toml`
 - [ ] Check unwrap, expect().
 - [ ] Check clone().
 - [ ] Check your commit messages.
 - [ ] Have you added meaningful comments.

### CODE REVIEW

File: crates/da-clients/aptos/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    async fn build_client(&self) -> AptosDaClient {
+        let client = Client::new(self.node_url.parse().unwrap());
+        let private_key =
+            Ed25519PrivateKey::from_encoded_string(&self.private_key).expect("Failed to parse private key");
+        let account_key = AccountKey::from(private_key);
+        let account_address = self.account_address.parse().expect("Invalid account address");
+        let account = LocalAccount::new(account_address, account_key, 500);
+        let module_address = self.module_address.parse().expect("Invalid module address");
+        let chain_id = ChainId::from_str(&self.chain_id).expect("Invalid chain id");
+        let trusted_setup = KzgSettings::load_trusted_setup_file(Path::new("./trusted_setup.txt"))
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/config.rs]: Make the path configurable

File: crates/da-clients/aptos/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        Self { chain_id, node_url, private_key, account_address, module_address }
+    }
+
+    async fn build_client(&self) -> AptosDaClient {
+        let client = Client::new(self.node_url.parse().unwrap());
+        let private_key =
+            Ed25519PrivateKey::from_encoded_string(&self.private_key).expect("Failed to parse private key");
+        let account_key = AccountKey::from(private_key);
+        let account_address = self.account_address.parse().expect("Invalid account address");
+        let account = LocalAccount::new(account_address, account_key, 500);
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/config.rs]:
```suggestion
        let account = LocalAccount::new(account_address, account_key, 0);
```

File: crates/da-clients/aptos/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+async fn prepare_blob(
+    state_diff: &[Vec<u8>],
+    trusted_setup: &KzgSettings,
+) -> color_eyre::Result<(Vec<FixedBytes<131072>>, Vec<FixedBytes<48>>, Vec<FixedBytes<48>>)> {
+    let mut sidecar_blobs = vec![];
+    let mut sidecar_commitments = vec![];
+    let mut sidecar_proofs = vec![];
+
+    for blob_data in state_diff {
+        let mut fixed_size_blob: [u8; BYTES_PER_BLOB] = [0; BYTES_PER_BLOB];
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]:
```suggestion
        let mut fixed_size_blob = [0; BYTES_PER_BLOB];
```

File: crates/da-clients/aptos/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        // Sign transaction.
+        let signed_txn = self.account.sign_transaction(txn);
+
+        // Submit transaction.
+        let tx = self
+            .client
+            .submit_and_wait(&signed_txn)
+            .await
+            .expect("Failed to submit update state transaction")
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]:
```suggestion
            ?
```

File: crates/da-clients/aptos/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+const STARKNET_VALIDITY: &str = "starknet_validity";
+const BLOB_SUBMISSION: &str = "blob_submission";
+
+#[async_trait]
+impl DaClient for AptosDaClient {
+    async fn publish_state_diff(&self, state_diff: Vec<Vec<u8>>, _to: &[u8; 32]) -> color_eyre::Result<String> {
+        let (blobs, commitments, proofs) = prepare_blob(&state_diff, &self.trusted_setup).await?;
+
+        let sequence_number =
+            self.client.get_account(self.account.address()).await.unwrap().into_inner().sequence_number;
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]:
```suggestion
            self.client.get_account(self.account.address()).await?.into_inner().sequence_number;
```

File: crates/da-clients/aptos/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        }
+    }
+
+    async fn max_blob_per_txn(&self) -> u64 {
+        // This number can be changed in the future, we will decide this value later.
+        6
+    }
+
+    async fn max_bytes_per_blob(&self) -> u64 {
+        // Similar with max_blob_per_txn
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]:
```suggestion
        // Similar to max_blob_per_txn
```

File: crates/da-clients/aptos/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
+use crate::AptosDaClient;
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/config.rs]: optimize imports

File: crates/utils/src/env_utils.rs [hunk 8]
```diff
@@ -5,7 +5,7 @@ pub fn get_env_var(key: &str) -> Result<String, VarError> {
 }
 
 pub fn get_env_var_or_panic(key: &str) -> String {
-    get_env_var(key).unwrap_or_else(|e| panic!("Failed to get env var {}: {}", key, e))
+    get_env_var(key).unwrap_or_else(|e|     panic!("Failed to get env var {}: {}", key, e))
```
Comment (zk-steve) [inline crates/utils/src/env_utils.rs]: ?

File: crates/da-clients/aptos/src/config.rs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+#[async_trait]
+impl DaConfig<AptosDaClient> for AptosDaConfig {
+    fn new_from_env() -> Self {
+        dotenv().expect("Failed to load .env file");
+        let node_url = get_env_var_or_panic("APTOS_NODE_URL");
+        let private_key = get_env_var_or_panic("APTOS_PRIVATE_KEY");
+        let account_address = get_env_var_or_panic("APTOS_ACCOUNT_ADDRESS");
+        let module_address = get_env_var_or_panic("APTOS_MODULE_ADDRESS");
+        let chain_id = get_env_var_or_panic("CHAIN_ID");
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/config.rs]:
```suggestion
        let chain_id = get_env_var_or_panic("APTOS_CHAIN_ID");
```

File: crates/da-clients/aptos/src/config.rs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        let trusted_setup = get_env_var_or_panic("TRUSTED_SETUP");
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/config.rs]:
```suggestion
        let trusted_setup = get_env_var_or_panic("APTOS_CRS_PATH");
```

File: crates/da-clients/aptos/src/helper.rs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+use aptos_sdk::transaction_builder::TransactionBuilder;
+use aptos_sdk::types::chain_id::ChainId;
+use aptos_sdk::types::transaction::{SignedTransaction, TransactionPayload};
+use aptos_sdk::types::LocalAccount;
+use std::time::SystemTime;
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/helper.rs]: optimize imports

File: crates/da-clients/aptos/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+const STARKNET_VALIDITY: &str = "starknet_validity";
+const BLOB_SUBMISSION: &str = "blob_submission";
+
+#[async_trait]
+impl DaClient for AptosDaClient {
+    async fn publish_state_diff(&self, state_diff: Vec<Vec<u8>>, _to: &[u8; 32]) -> color_eyre::Result<String> {
+        // Create blobs, commitments, proofs from state_diff and trusted setup and transfer to "MoveValue".
+        let data = prepare_blob(&state_diff, &self.trusted_setup).await?;
+        let loop_cycle = get_env_var_or_panic("LOOP_CYCLE").parse::<usize>()?;
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]: Remove this

File: crates/da-clients/aptos/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+            txs.push(tx);
+        }
+
+        // Submit transaction.
+        let pending_transactions = txs
+            .into_iter()
+            .map(|tx| {
+                let client = self.client.clone();
+                tokio::spawn(async move {
+
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]: remove empty lines

File: crates/da-clients/aptos/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        }
+
+        // Submit transaction.
+        let pending_transactions = txs
+            .into_iter()
+            .map(|tx| {
+                let client = self.client.clone();
+                tokio::spawn(async move {
+                    
+                    client.submit(&tx).await.unwrap().into_inner()
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]: handle error instead of unwraping it

File: crates/da-clients/aptos/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+            results.push(handle.await.unwrap());
+        }
+
+        // Wait for transaction.
+        let results = results
+            .into_iter()
+            .map(|pending_tx| {
+                let client = self.client.clone();
+                tokio::spawn(async move {
+                    let transaction = client.wait_for_transaction(&pending_tx).await.unwrap().into_inner();
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs:107]: handle error

File: Cargo.toml [hunk 16]
```diff
@@ -88,3 +90,9 @@ prover-client-interface = { path = "crates/prover-services/prover-client-interfa
 gps-fact-checker = { path = "crates/prover-services/gps-fact-checker" }
 sharp-service = { path = "crates/prover-services/sharp-service" }
 orchestrator = { path = "crates/orchestrator" }
+
+# Aptos
+aptos-sdk = { git = "https://github.com/sota-zk-labs/aptos-core" }
+[patch.crates-io]
+merlin = { git = "https://github.com/aptos-labs/merlin" }
+x25519-dalek = { git = "https://github.com/Tranduy1dol/x25519-dalek", branch = "zeroize_v1.7" }
```
Comment (zk-steve) [inline Cargo.toml:107]:
```suggestion
x25519-dalek = { git = "https://github.com/aptos-labs/x25519-dalek", branch = "zeroize_v1" }
```

File: crates/da-clients/aptos/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+            });
+
+        // Build and sign transaction
+        let sequence_number = self.client.get_account(self.account.address()).await?.into_inner().sequence_number;
+        self.account.set_sequence_number(sequence_number);
+        let mut txs = vec![];
+
+        for i in 0..BYTES_PER_BLOB / MAX_BLOB_PART_SIZE {
+            let payload = TransactionPayload::EntryFunction(EntryFunction::new(
+                ModuleId::new(self.account.address(), Identifier::new(STARKNET_VALIDITY).unwrap()),
```
Comment (0x5ea000000) [inline crates/da-clients/aptos/src/lib.rs]: take out of the loop

File: crates/da-clients/aptos/src/lib.rs (hunk 15 above)
Comment (0x5ea000000) [reply crates/da-clients/aptos/src/lib.rs:107]: @Tranduy1dol r u sure you resolved this?

File: crates/da-clients/aptos/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+
+    async fn verify_inclusion(&self, external_id: &str) -> color_eyre::Result<DaVerificationStatus> {
+        let client = &self.client;
+
+        let hash_split = external_id.split("0x").filter(|&s| !s.is_empty()).collect::<Vec<_>>();
+        let failed_hashes: Vec<&str> = Vec::new();
+
+        for hash_ptr in hash_split {
+            let hash = HashValue::from_str(hash_ptr)?;
+            match client.get_transaction_by_hash(hash).await {
```
Comment (0x5ea000000) [inline crates/da-clients/aptos/src/lib.rs]: consider using tokio spawn to parallel checking the hashes

File: crates/da-clients/aptos/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+        if container.is_none() {
+            let aptos_container = AptosContainer::init().await.unwrap();
+
+            aptos_container.faucet(&MODULE_ACCOUNT).await.unwrap();
+            aptos_container.faucet(&SENDER_ACCOUNT).await.unwrap();
+
+            let mut named_addresses = HashMap::new();
+            named_addresses.insert("starknet_addr".to_string(), MODULE_ACCOUNT.address().to_string());
+
+            aptos_container.upload_contract("../../../../ionia", &MODULE_ACCOUNT, &named_addresses).await.unwrap();
```
Comment (0x5ea000000) [inline crates/da-clients/aptos/src/lib.rs]: consider adding ionia to git submodule

File: crates/da-clients/aptos/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+        let da_client = init_aptos_client().await;
+
+        let chain_id = da_client.chain_id.clone();
+        let module_address = da_client.module_address.clone();
+
+        let sequence_number =
+            da_client.client.get_account(da_client.account.address()).await.unwrap().into_inner().sequence_number;
+        da_client.account.set_sequence_number(sequence_number);
+
+        let sequencer_number = da_client.account.increment_sequence_number();
```
Comment (0x5ea000000) [inline crates/da-clients/aptos/src/lib.rs]: reassign sequence_number?

File: crates/da-clients/aptos/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+
+        let tx_builder = TransactionBuilder::new(
+            TransactionPayload::EntryFunction(EntryFunction::new(
+                ModuleId::new(module_address, Identifier::new(STARKNET_VALIDITY).unwrap()),
+                Identifier::new(INIT_CONTRACT_STATE).unwrap(),
+                vec![],
+                serialize_values(
+                    vec![
+                        &MoveValue::U256(u256::U256::from_str_radix("0", 10).unwrap()),
```
Comment (0x5ea000000) [inline crates/da-clients/aptos/src/lib.rs]: consider writing a helper function to convert str to move value u256

**Review by 0x5ea000000 — COMMENTED** [review]
check the comment

File: bash.sh
Comment (zk-steve) [inline bash.sh:1]: remove this file

File: crates/da-clients/aptos/src/helper.rs [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+) -> SignedTransaction {
+    let i = sender.increment_sequence_number();
+    let tx = TransactionBuilder::new(
+        payload,
+        SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs() + 60,
+        chain_id,
+    )
+    .sender(sender.address())
+    .sequence_number(i)
+    .max_gas_amount(100000)
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/helper.rs]: 30000

File: crates/da-clients/aptos/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+
+pub struct AptosDaClient {
+    pub client: Client,
+    pub account: LocalAccount,
+    pub module_address: AccountAddress,
+    pub chain_id: ChainId,
+    pub trusted_setup: KzgSettings,
+}
+
+const STARKNET_VALIDITY: &str = "starknet_validity";
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]: consts above structs

File: .github/workflows/linters-cargo.yml [hunk 24]
```diff
@@ -18,6 +18,7 @@ jobs:
       - uses: Swatinem/rust-cache@v2
       - name: Format and clippy
         run: |
-          cargo fmt -- --check
-          cargo clippy --no-deps -- -D warnings
-          cargo clippy --tests --no-deps -- -D warnings
+          sudo apt install build-essential cmake pkg-config libudev-dev
```
Comment (zk-steve) [inline .github/workflows/linters-cargo.yml:21]: Is it necessary to do changes like this?

File: crates/da-clients/aptos/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+            })
+            .fold((vec![], vec![], vec![]), |(mut blobs, mut commitments, mut proofs), (blob, commitment, proof)| {
+                blobs.push(blob);
+                commitments.push(commitment);
+                proofs.push(proof);
+                (blobs, commitments, proofs)
+            });
+
+        // Build and sign transaction
+        let sequencer_number = self.client.get_account(self.account.address()).await?.into_inner().sequence_number;
```
Comment (zk-steve) [inline crates/da-clients/aptos/src/lib.rs]: sequencer_number ?

File: CHANGELOG.md [hunk 26]
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 - `AWS_DEFAULT_REGION="localhost"` var. in .env.test for omniqueue queue testing.
 - Added basic rust-toolchain support.
 - Tests for DA job.
+- Aptos DA client, Aptos-compatible workflow.
```
Comment (zk-steve) [inline CHANGELOG.md:25]: https://keepachangelog.com/en/1.1.0/

File: .github/workflows/pull-request.yml [hunk 27]
```diff
@@ -4,9 +4,9 @@ name: Workflow - Pull Request
 on:
   workflow_dispatch:
   pull_request:
-    branches: [main]
+    branches: [aptos]
   push:
-    branches: [main]
+    branches: [aptos]
```
Comment (zk-steve) [inline .github/workflows/pull-request.yml:9]: remove

**Review by zk-steve — APPROVED** [review]

**Pull request merged**